src/gui/creds/webflowcredentials.cpp

```diff
@@ -0,0 +1,252 @@
+#include <QLabel>
+
+#include "accessmanager.h"
+#include "account.h"
+#include "theme.h"
+#include "wizard/webview.h"
+#include "webflowcredentialsdialog.h"
+
+using namespace QKeychain;
+
+namespace OCC {
+
+Q_LOGGING_CATEGORY(lcWebFlowCredentials, "sync.credentials.webflow", QtInfoMsg)
+
+WebFlowCredentials::WebFlowCredentials()
+    : _ready(false),
+      _credentialsValid(false)
+{
+
+}
+
+WebFlowCredentials::WebFlowCredentials(const QString &user, const QString &password, const QSslCertificate &certificate, const QSslKey &key)
+    : _user(user)
+    , _password(password)
+    , _clientSslKey(key)
+    , _clientSslCertificate(certificate)
+    , _ready(true)
+    , _credentialsValid(true)
+{
+
+}
+
+QString WebFlowCredentials::authType() const {
+    return QString::fromLatin1("webflow");
+}
+
+QString WebFlowCredentials::user() const {
+    return _user;
+}
+
+QString WebFlowCredentials::password() const {
+    return _password;
+}
+
+QNetworkAccessManager *WebFlowCredentials::createQNAM() const {
+    qCInfo(lcWebFlowCredentials()) << "Get QNAM";
+    AccessManager *qnam = new AccessManager();
+
+    connect(qnam, &AccessManager::authenticationRequired, this, &WebFlowCredentials::slotAuthentication);
+    connect(qnam, &AccessManager::finished, this, &WebFlowCredentials::slotFinished);
+
+    return qnam;
+}
+
+bool WebFlowCredentials::ready() const {
+    return _ready;
+}
+
+void WebFlowCredentials::fetchFromKeychain() {
+    _wasFetched = true;
+
+    // Make sure we get the user fromt he config file
+    fetchUser();
+
+    if (ready()) {
+        emit fetched();
+    } else {
+        qCInfo(lcWebFlowCredentials()) << "Fetch from keyhchain!";
+        fetchFromKeychainHelper();
+    }
+}
+
+void WebFlowCredentials::askFromUser() {
+    _askDialog = new WebFlowCredentialsDialog();
+
+    QUrl url = _account->url();
+    QString path = url.path() + "/index.php/login/flow";
+    url.setPath(path);
+    _askDialog->setUrl(url);
+
+    QString msg = tr("You have been logged out of %1 as user %2. Please login again")
+            .arg(_account->displayName(), _user);
+    _askDialog->setInfo(msg);
+
+    _askDialog->show();
+
+    connect(_askDialog, &WebFlowCredentialsDialog::urlCatched, this, &WebFlowCredentials::slotAskFromUserCredentialsProvided);
+
+    qCWarning(lcWebFlowCredentials()) << "User needs to reauth!";
+}
+
+void WebFlowCredentials::slotAskFromUserCredentialsProvided(const QString &user, const QString &pass, const QString &host) {
+    if (_user != user) {
+        qCInfo(lcWebFlowCredentials()) << "Authed with the wrong user!";
+
+        QString msg = tr("Please login with the user: %1")
+                .arg(_user);
+        _askDialog->setError(msg);
+
+        QUrl url = _account->url();
+        QString path = url.path() + "/index.php/login/flow";
+        url.setPath(path);
+        _askDialog->setUrl(url);
+
+        return;
+    }
+
+    qCInfo(lcWebFlowCredentials()) << "New password is:" << pass;
```

Comment:
Not sure if we should log passwords at info level.

Reply:
Ah yes debug code. I'll fix that all up :)